### PR TITLE
automation: wait_for allow custom output while waiting

### PR DESCRIPTION
**Pull request by Itxaka**
For some long commands it would be nice to get some extra output
while they are running. This allows to pass a 6th arg to wait_for
that would get evaluated instead of the usual dots.

### CODE REVIEW

**Comment by Itxaka** [comment]
[Make the mkcloud more verbose](https://trello.com/c/iVyd9C60/293-make-the-mkcloud-more-verbose)

**Comment by Itxaka** [comment]
@bmwiedemann and @abel-navarro I know you guys are awesome with bash, maybe you can see some improvements or fallbacks of this approach? Any hints?

**Comment by jsuchome** [comment]
Could you already use it in some place (e.g. when doing `crowbarctl upgrade ...` calls)?

**Comment by Itxaka** [comment]
@jsuchome if you prefer yes, but was waiting for the bash masters to tell me how can I improve it, before jumping into it :D

**Comment by jsuchome** [comment]
I think the example might be convincing :-)

**Comment by Itxaka** [comment]
@jsuchome done!

```
root@crowbar:~ # wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish" "complain 15 Nodes step has failed" "printf 'Nodes left to upgrade: %s' `crowbarctl upgrade status nodes --plain |grep not_upgraded|cut -d ' ' -f 2|wc -l` "
Waiting for: 'nodes' upgrade step to finish
  until this condition is true: crowbar_nodeupgrade_finished
  waiting 360 cycles of 30 seconds = 10800 seconds
Nodes left to upgrade: 5

```

**Comment by Itxaka** [comment]
I dont like it that much tbh. Depending on terminal and font size and number of nodes from the output it could lead to line wrapping and then the carriage return would not go back to where its supposed to, making the output unreadable :/

New output:
```
root@crowbar:~ # wait_for 360 3 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish" "complain 15 Nodes step has failed" "crowbarctl upgrade status nodes --plain |grep not_upgraded|cut -d ' ' -f 2 |cut -d '.' -f 1|sort|tr '\n' ','| sed 's/,$//'|xargs printf 'Upgrading nodes: %s\r'"
Waiting for: 'nodes' upgrade step to finish
  until this condition is true: crowbar_nodeupgrade_finished
  waiting 360 cycles of 3 seconds = 1080 seconds
Upgrading nodes: d52-54-77-77-77-01,d52-54-77-77-77-02,d52-54-77-77-77-03,d52-54-77-77-77-04,d52-54-77-77-77-05```

**Comment by jsuchome** [comment]
That long line you are creating is really horrible :-). So let's see what others think about it.

But I was actually thinking about parsing `current_nodes` value from the output of `crowbarctl upgrade status`, not `crowbarctl upgrade status nodes`. Or even parsing the status file directly (`$upgrade_progress_file`).

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -5130,7 +5130,7 @@ function onadmin_crowbar_nodeupgrade
             wait_for 300 5 "grep current_step $upgrade_progress_file | grep -v backup_openstack" "backup openstack step to finish"
 
             safely crowbarctl upgrade nodes all
-            wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish"
+            wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish" "'Nodes' step has failed. Check the upgrade status." "crowbarctl upgrade status nodes --plain |grep not_upgraded|cut -d ' ' -f 2 |cut -d '.' -f 1|sort|tr '\n' ','| sed 's/,$//'|xargs printf 'Upgrading nodes: %s\r'"
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: your errorcmd is missing an echo and an exit

**Comment by bmwiedemann** [comment]
I already cherry-picked the first commit to master because it is nice+simple+working.

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (Itxaka) [inline scripts/qa_crowbarsetup.sh]: :+1:

**Comment by bmwiedemann** [comment]
I think, you could replace much of the many pipes with a well-crafted sed call.
e.g. `echo -e 'foo\nbar' | sed -n /foo/{s/o/x/p}` only looks at lines matching a pattern and then does some **s**ubstitutions on it before **p**rinting
but in the end, it remains a matter of taste.

**Comment by Itxaka** [comment]
upgrade failure is due to

```
steps.prepare.errors.prepare.data | undefined method `chef_data_bag_destroy' for #<ProvisionerService:0x00000008720cb0> |
| steps.prepare.errors.prepare.help | Check /var/log/crowbar/production.log at admin server.      
```

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -5130,7 +5130,7 @@ function onadmin_crowbar_nodeupgrade
             wait_for 300 5 "grep current_step $upgrade_progress_file | grep -v backup_openstack" "backup openstack step to finish"
 
             safely crowbarctl upgrade nodes all
-            wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish"
+            wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish" "complain 13 'Nodes step has failed. Check the upgrade status.'" "crowbarctl upgrade status nodes --plain |grep not_upgraded|cut -d ' ' -f 2 |cut -d '.' -f 1|sort|tr '\n' ','| sed 's/,$//'|xargs printf 'Upgrading nodes: %s\r'"
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: 1.
The value for `$error_cmd` is redundant. The `wait_for` function already prints a useful message. If its not useful then rather adapt the `waitfor` string. So my proposal is to pass an empty string here.

2.
The print while is a nice idea. If you add looong stuff as value though, it is a hint that you instead wanted a function and then rather just pass as function name as print_while string.

So I recommend to do this:

```
function show_crowbar_upgrade_status
{
    crowbarctl upgrade status nodes --plain | grep not_upgraded | cut -d ' ' -f 2 | cut -d '.' -f 1 | \
        sort | tr '\n' ',' | sed 's/,$//' | xargs printf 'Upgrading nodes: %s\r'
}

[...]

wait_for 360 30 "crowbar_nodeupgrade_finished" "'nodes' upgrade step to finish" "" "show_crowbar_upgrade_status"
```

Comment (Itxaka) [inline scripts/qa_crowbarsetup.sh]: 1. -> I thought this was useful for tracking the errors, having it exit with a complain XX would probably be useful for something, maybe the jenkins log parser?
2. -> :+1:

Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: Here are [the current log parser rules](https://github.com/SUSE-Cloud/automation/blob/master/scripts/jenkins/log-parser/openstack-mkcloud-rules.txt); enhancements extremely welcome!

Comment (Itxaka) [inline scripts/qa_crowbarsetup.sh]: Looks like using `complain` is already covered in the parser rules, as in the end it exists with `Error XX` which is properly parsed.

**Comment by Itxaka** [comment]
@jdsn updated with your recommendations!

**Pull request merged**